src/message/messagesReducer.js

```diff
@@ -15,126 +16,110 @@ import {
   EVENT_REACTION_REMOVE,
   EVENT_UPDATE_MESSAGE,
 } from '../actionConstants';
-import { NULL_ARRAY, NULL_OBJECT } from '../nullObjects';
-import { groupItemsById } from '../utils/misc';
+import { NULL_ARRAY } from '../nullObjects';
 
-const initialState: MessagesState = NULL_OBJECT;
+const initialState: MessagesState = Immutable.Map([]);
 
 const eventReactionAdd = (state, action) => {
-  const oldMessage = state[action.message_id];
+  const oldMessage = state.get(action.message_id);
   if (!oldMessage) {
     return state;
   }
-  return {
-    ...state,
-    [action.message_id]: {
-      ...oldMessage,
-      reactions: oldMessage.reactions.concat({
-        emoji_name: action.emoji_name,
-        user_id: action.user_id,
-        reaction_type: action.reaction_type,
-        emoji_code: action.emoji_code,
-      }),
-    },
-  };
+  return state.set(action.message_id, {
+    ...oldMessage,
+    reactions: oldMessage.reactions.concat({
+      emoji_name: action.emoji_name,
+      user_id: action.user_id,
+      reaction_type: action.reaction_type,
+      emoji_code: action.emoji_code,
+    }),
+  });
```

Comment:
This pattern can be condensed like so (in a followup), using [`Immutable.map#update`](https://immutable-js.github.io/immutable-js/docs/#/Map/update):
```js
const eventReactionAdd = (state, action) =>
  state.update(
    action.message_id,
    oldMessage =>
      oldMessage && {
        ...oldMessage,
        reactions: oldMessage.reactions.concat({
          emoji_name: action.emoji_name,
          user_id: action.user_id,
          reaction_type: action.reaction_type,
          emoji_code: action.emoji_code,
        }),
      },
  );
```

A crucial feature here of `update` is that
> if the `updater` function returns the same value it was called with, then no change will occur.

So if the message isn't present, the updater will be passed `undefined`, and just has to return `undefined` and the map will be returned unchanged.

I'm kind of ambivalent on this clarity-wise. It's a bit shorter, but does rely on a somewhat subtle fact about `update`'s interface. But two things push me toward thinking that overall it's an improvement:
* It should be more efficient: the old code requires the key to be looked up in the data structure twice, and with this version Immutable can (and I expect it does) look it up just once.
* Because it no longer needs a local variable, it can be inlined conveniently into the `switch` in the reducer. With similar simplifications, all of these helpers can be inlined conveniently into the reducer. That's the style of most of our reducers, and I generally prefer it because it means one fewer hop to look up what the effect of an action is (or vice versa, to find in what situation some code found in one of these is called.)

Those and:
* If we start making use of this feature routinely, as we well might, then it becomes a familiar one :slightly_smiling_face: and not a puzzling subtlety.